--- a/sound/soc/rockchip/rockchip_i2s_tdm.c
+++ b/sound/soc/rockchip/rockchip_i2s_tdm.c
@@ -586,6 +586,8 @@
     udelay(10);
 }
 
+static void rockchip_i2s_tdm_post_stop_reset(struct rk_i2s_tdm_dev *i2s_tdm);
+
 static int rockchip_i2s_tdm_clear(struct rk_i2s_tdm_dev *i2s_tdm,
       unsigned int clr)
 {
@@ -611,7 +613,9 @@
     ret = regmap_read_poll_timeout_atomic(i2s_tdm->regmap, I2S_CLR, val,
               !(val & clr), 10, 100);
     if (ret < 0) {
-    dev_warn(i2s_tdm->dev, "failed to clear %u\n", clr);
+    dev_warn(i2s_tdm->dev, "failed to clear %u; real reset #1\n", clr);
+    rockchip_i2s_tdm_post_stop_reset(i2s_tdm);
+    msleep(200);
     goto reset;
     }
 
